### PR TITLE
initial support for archive storage spaces

### LINKED ISSUE #588: space layout to support data and index zng files

### PR DESCRIPTION
**Pull request by alfred-landrum**
* Move the existing space support into its own package, `unizng`. Log and pcap import only work against unizng.
* Add a Storage interface and a storage loader for storage kind detection.
* Add the archive storage package that supports zql searches.
* Add a test that creates a space whose data_path points to an existing archive directory, and verify that zql searching works.

I've also manually setup an archive space with zqd, and verified that Brim zql searches work as expected.

Closes #588

### CODE REVIEW

File: zbuf/combiner.go [hunk 1]
```diff
@@ -4,25 +4,45 @@ import (
 	"fmt"
 	"io"
 
+	"github.com/brimsec/zq/pkg/nano"
 	"github.com/brimsec/zq/zng"
 )
 
+type Direction bool
```
Comment (alfred-landrum) [inline zbuf/combiner.go:11]: @mattnibs : I know an early version of your Storage PR had a similar direction option for Combiner, but I didn't have a copy of that, so not sure how this compares with your implementation, PTAL.

File: zqd/handlers.go [hunk 2]
```diff
@@ -159,7 +159,8 @@ func handlePcapSearch(c *Core, w http.ResponseWriter, r *http.Request) {
 }
 
 func handleSpaceList(c *Core, w http.ResponseWriter, r *http.Request) {
-	spaces, err := c.spaces.List()
+	ctx := r.Context()
+	spaces, err := c.spaces.List(ctx)
```
Comment (nwt) [inline zqd/handlers.go:163]:
```suggestion
	spaces, err := c.spaces.List(r.Context())
```

File: zqd/handlers.go [hunk 3]
```diff
@@ -195,12 +196,13 @@ func handleSpacePost(c *Core, w http.ResponseWriter, r *http.Request) {
...
 	}
 
+	ctx := r.Context()
 	sp, err := c.spaces.Create(req.Name, req.DataPath)
 	if err != nil {
 		respondError(c, w, r, err)
 		return
 	}
-	info, err := sp.Info()
+	info, err := sp.Info(ctx)
```
Comment (nwt) [inline zqd/handlers.go:205]:
```suggestion
	info, err := sp.Info(r.Context())
```

File: zqd/storage/archive/archive.go [hunk 4]
```diff
@@ -0,0 +1,81 @@
+package archivestore
```
Comment (nwt) [inline zqd/storage/archive/archive.go:1]: The mismatch between directory and package name here is surprising. Why not change the directory name to archivestore?

File: zqd/storage/storage.go [hunk 5]
```diff
@@ -0,0 +1,41 @@
...
+	"github.com/brimsec/zq/pkg/nano"
+	"github.com/brimsec/zq/zbuf"
+)
+
+type Kind int
+
+const (
+	KindUnknown Kind = iota
+	KindUniZng
+	KindArchive
```
Comment (nwt) [inline zqd/storage/storage.go:15]: How about KindZngFile and KindZngArchive?

File: zbuf/combiner.go (hunk 1 above)
Comment (mattnibs) [reply zbuf/combiner.go:11]: My version used expr.SortFn which IMO makes this more flexible. You can see my pass for Combiner here: https://github.com/brimsec/zq/pull/823/files#diff-e97e96d1cbbd75b1be023a168075c8ec

Comment (alfred-landrum) [reply zbuf/combiner.go]: updated in 5dd8d23

**Pull request merged**